std-bits/table/src/main/java/org/enso/table/data/table/join/LookupJoin.java

```diff
@@ -0,0 +1,205 @@
+package org.enso.table.data.table.join;
+
+import org.enso.base.text.TextFoldingStrategy;
+import org.enso.table.data.column.builder.Builder;
+import org.enso.table.data.column.storage.Storage;
+import org.enso.table.data.index.Index;
+import org.enso.table.data.index.MultiValueIndex;
+import org.enso.table.data.index.UnorderedMultiValueKey;
+import org.enso.table.data.mask.OrderMask;
+import org.enso.table.data.table.Column;
+import org.enso.table.data.table.Table;
+import org.enso.table.error.NonUniqueLookupKey;
+import org.enso.table.error.NullValuesInKeyColumns;
+import org.enso.table.error.UnmatchedRow;
+import org.enso.table.util.ConstantList;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.IntStream;
+
+public class LookupJoin {
+  private static final TextFoldingStrategy TEXT_FOLDING = TextFoldingStrategy.unicodeNormalizedFold;
+
+  public static Table lookupAndReplace(List<Equals> keys, List<LookupColumnDescription> columnDescriptions,
+                                       boolean allowUnmatchedRows) {
+    if (keys.isEmpty()) {
+      throw new IllegalArgumentException("No join keys specified.");
+    }
+
+    LookupJoin joiner = new LookupJoin(keys, columnDescriptions, allowUnmatchedRows);
+    joiner.verifyLookupUniqueness();
+    return joiner.join();
+  }
+
+  private final List<LookupColumnDescription> columnDescriptions;
+  private final List<String> keyColumnNames;
+
+  private final MultiValueIndex<UnorderedMultiValueKey> lookupIndex;
+
+  private final Storage<?>[] baseKeyStorages;
+  private final List<TextFoldingStrategy> textFoldingStrategies;
+  private final int baseTableRowCount;
+  private final boolean allowUnmatchedRows;
+
+  private UnorderedMultiValueKey makeTableRowKey(int ix) {
+    return new UnorderedMultiValueKey(baseKeyStorages, ix, textFoldingStrategies);
+  }
+
+  private LookupJoin(List<Equals> keys, List<LookupColumnDescription> columnDescriptions, boolean allowUnmatchedRows) {
+    baseKeyStorages = keys.stream().map(Equals::left).map(Column::getStorage).toArray(Storage[]::new);
+    this.columnDescriptions = columnDescriptions;
+    this.allowUnmatchedRows = allowUnmatchedRows;
+    textFoldingStrategies = ConstantList.make(TEXT_FOLDING, baseKeyStorages.length);
+
+    Column[] lookupKeyColumns = keys.stream().map(Equals::right).toArray(Column[]::new);
+    checkNullsInKey(lookupKeyColumns);
+    keyColumnNames = Arrays.stream(lookupKeyColumns).map(Column::getName).toList();
+
+    assert lookupKeyColumns.length > 0;
+    // tableSize parameter is only needed if there are no key columns, but that is not possible
+    lookupIndex = MultiValueIndex.makeUnorderedIndex(lookupKeyColumns, 0, textFoldingStrategies);
+    baseTableRowCount = baseKeyStorages[0].size();
+  }
+
+  private void verifyLookupUniqueness() {
+    for (Map.Entry<UnorderedMultiValueKey, List<Integer>> group : lookupIndex.mapping().entrySet()) {
```

Comment:
why not carry a flag in `MultiValueIndex` of `isUnique`? Any time a key refers to more than 1 we can set to false.